_posts/2023-09-19-virtual-thread-1.adoc

```diff
@@ -0,0 +1,332 @@
+Increasing the thread pool size and, consequently, increasing the memory usage can blow up your Cloud bill and deployment density.
+
+The reactive movement proposed an alternative model to work around that issue. 
+It promotes the usage of non-blocking I/O and asynchronous development models to use resources (CPU and memory) more efficiently.
+With the reactive model, **a single thread can handle multiple concurrent requests**. 
+So, instead of having a large pool of threads, you have a minimum number of threads (generally equal to the number of CPU cores).
+This small amount of threads, often named event loop, handles all your requests.
+When a request is received, it calls the processing code on one of these threads. 
+When the processing needs to execute an I/O operation, instead of using blocking I/O, it schedules the operations and passes a **continuation**.
+This continuation is the code to be invoked when the I/O completes, so basically, the rest of the processing. 
+The reactive model is highly efficient. 
+Projects like Netty or Vertx handle the heavy lifting at the I/O layer, allowing the users to _just_ write the processing code.
+But there is a catch. 
+As mentioned, you need to write your code as a chain of continuations. 
+While there are multiple approaches, such as callbacks, futures, reactive programming, or co-routines, it always complicates the code, especially how you reason about it.
+It's not the only issue. 
+Your code can not only block during I/O operation; it must not execute lengthy processing (what we call monopolization).
+Indeed, the model's efficiency comes from the ability to process many concurrent requests concurrently. 
+If the thread is used for a long time, it does not allow the other requests to be processed, and, as for the imperative model,  you start piling up requests.
+
+For example, the following snippet is equivalent to the previous one: it receives an HTTP request, calls another HTTP service, and returns a response. But this time, it uses the Quarkus reactive model.
+
+[source, java]
+----
+@Path("/greetings")
+public class ReactiveApp {
+
+  @RestClient RemoteService service;
+
+  @GET
+  public Uni<String> process() {
+    // Runs on an event loop (platform) thread because the
+    // method uses a asynchronous signature
+
+    // `service` is a rest client, it executes an I/O operation
+    // but this time it returns a Uni<String>, so it's not blocking
+    return service.asyncGreetings() // Non-blocking
+        .map(resp -> {
+          // This is the continuation, the code executed once the
+          // response is received
+          return resp.toUpperCase();
+        });
+    }
+}
+----
+
+Most of the time, the reactive and imperative models are opposed.
+This is not the case. 
```

Comment:
this reads weird - which "this" is it referring to ? should it have been "This does not need to be the case" ?